# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: Châu Tấn Kiệt
- MSSV: 21127329
- Lớp: 21CLC4

# Import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn import linear_model
from sklearn.model_selection import KFold

# Import thêm dữ thư viện nếu cần

# Đọc dữ liệu

In [2]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
x_train = train.iloc[:, :-1]   # Dataframe (chứa 10 đặc trưng huấn luyện)
y_train = train.iloc[:, -1]      # Series    (chứa 1 giá trị mục tiêu kiểm tra)

x_test = test.iloc[:, :-1]       # Dataframe (chứa 10 đặc trưng kiểm tra)
y_test = test.iloc[:, -1]        # Series    (chứa 1 giá trị mục tiêu kiểm tra)

# Sinh viên có thể sử dụng các khác nếu cần

# Cài đặt hàm

In [3]:
## Cài đặt các hàm cần thiết ở đây

## Tính MAE
def MAE(y_true, predictions):
    y_true, predictions = np.array(y_true), np.array(predictions)
    return np.mean(np.abs(y_true - predictions))


def LinearRegression(A, b):
    reg = linear_model.LinearRegression().fit(A, b)
    return reg.coef_, reg.intercept_

## Công thức hồi quy tuyến tính
def findFormula(reg):
    print('Intercept: \n', reg.intercept_)
    print('Coefficients: \n', reg.coef_)

## Tính cross-validation bằng K-fold
def CrossValidation(x, y, kfold):
    kf = KFold(n_splits=kfold)
    error_list = []

    for train_index, test_index in kf.split(x):
        x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]

        new_x, bias =  linear_model.LinearRegression().fit(x_train, y_train)
        error_list.append(MAE(y_test,np.dot(x_test, new_x) + bias))

    return np.mean(error_list)


def chooseBestProperty(A, b):
    result = []

    for i in range(A.shape[1]):
        cv_prop = CrossValidation(A[:, i:i + 1], b, 10)
        result.append(cv_prop)
    
    return result, np.argmin(result)


# Yêu cầu 1a: Sử dụng toàn bộ 11 đặc trưng đầu tiên `Gender`, `10percentage`, `12percentage`, `CollegeTier`, `Degree`, `collegeGPA`, `CollegeCityTier`, `English`, `Logical`, `Quant`, `Domain` (2 điểm) 

In [4]:
## Phần code cho yêu cầu 1a
x1  = x_train[['Gender', '10percentage','12percentage', 'CollegeTier','Degree', 'collegeGPA','CollegeCityTier', 'English','Logical', 'Quant','Domain']]
y1 = y_train
reg = linear_model.LinearRegression().fit(x1,y1)


In [21]:
x2 = x_test[['Gender', '10percentage','12percentage', 'CollegeTier','Degree', 'collegeGPA','CollegeCityTier', 'English','Logical','Quant','Domain']]
y2_pred = reg.predict(x2)
y2_pred

array([194207.93160491, 340719.58717406, 325416.84861397, 273672.74799754,
       298369.36726444, 352015.74694404, 234779.7619092 , 261961.54605652,
       266112.40481144, 372849.64049772, 301599.21406001, 279946.54486441,
       228413.01407561, 305770.45016172, 403117.84794154, 311514.75295673,
       255771.64061019, 249511.47496086, 282050.50784367, 342932.68002501,
       252047.78684692, 331783.85435779, 392592.81848187, 441556.83514414,
       237549.86845078, 406144.58635159, 325347.97152288, 340287.86083972,
       339589.96608669, 359034.7906052 , 337830.75631687, 388921.75950567,
       307399.94344878, 507243.30753988, 303046.98974608, 288503.42897022,
       488343.13883449, 307550.03410941, 373338.98609063, 281018.74597146,
       499929.31989582, 360613.09095212, 266862.5060087 , 386429.00791345,
       236977.45662334, 391749.09033547, 312501.27630642, 255610.43006438,
       474422.70383531, 300598.05757002, 293416.22841846, 369686.76439215,
       250401.61675547, 2

In [22]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
MAE(y_test, y2_pred)

105052.52978823145

In [23]:
findFormula(reg)

Intercept: 
 49248.089734813664
Coefficients: 
 [-23183.32950765    702.76679172   1259.0187879  -99570.60814074
  18369.9624496    1297.53200035  -8836.727123      141.75993906
    145.74234652    114.64331342  34955.75040521]


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = 49248.090 -23183.330{b_1} + 702.767{b_2} +  1259.018{b_3} - 99570.608{b_4} +
  18369.962{b_5} + 1297.532{b_6} -8836.727{b_7} + 141.760{b_8} + 145.742{b_9} +114.643{b_10} + 34955.750{b_11}$$

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng tính cách với các đặc trưng tính cách gồm `conscientiousness`, `agreeableness`, `extraversion`, `nueroticism`, `openess_to_experience`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [20]:
x_train_b = train[['conscientiousness','agreeableness', 'extraversion', 'nueroticism', 'openess_to_experience']].to_numpy()
print(x_train_b)
error_list, best_index = chooseBestProperty(x_train_b, y_train.to_numpy())

print(error_list)
# Xây dựng mô hình
x_b, bias_b = LinearRegression(x_train_b[:, best_index:best_index + 1], y_train.to_numpy())


[[-0.159    0.3789   1.2396   0.1459   0.2889 ]
 [ 1.1336   0.0459   1.2396   0.5262  -0.2859 ]
 [ 0.51    -0.1232   1.5428  -0.2902  -0.2875 ]
 ...
 [-0.1082   0.0328  -0.4891  -0.2902   0.5024 ]
 [-0.881    0.1888  -0.344    0.0623   0.6603 ]
 [ 1.4374   1.2808  -0.4891  -1.46537  0.5419 ]]


TypeError: cannot unpack non-iterable LinearRegression object

In [12]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

x3,y3 = LinearRegression(x_train[['conscientiousness','agreeableness', 'extraversion', 'nueroticism', 'openess_to_experience']].to_numpy(), y_train.to_numpy())
error_list, best_index = chooseBestProperty(x3,y3)

print(error_list)
# Xây dựng mô hình
x_b, bias_b = LinearRegression(x_train[:, best_index:best_index + 1], y_train)

IndexError: tuple index out of range

In [ ]:
# Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện

In [ ]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$

# Yêu cầu 1c: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng `English`, `Logical`, `Quant`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [ ]:
# Phần code cho yêu cầu 1c
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu
x5 = x_train[['English','Logical', 'Quant']].to_numpy()
y5 = y_train

In [ ]:
# Huấn luyện lại mô hình best_skill_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện

In [ ]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_skill_feature_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$

# Yêu cầu 1d: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

## Tìm mô hình

In [ ]:
# Trình bày các phần tìm ra mô hình

## Thử nghiệm, so sánh các mô hình

In [ ]:
# Phần code cho yêu cầu 1d
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu

In [ ]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện

In [ ]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model

Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = ...$$